x-pack/plugins/cloud_security_posture/public/pages/rules/rules_table_header.tsx

```diff
@@ -144,14 +112,15 @@ export const RulesTableHeader = ({
                   }
                 )}
                 id={'rule-number-multi-select-filter'}
-                onChange={(ruleNumber) => {
-                  setSelectedRuleNumber([...ruleNumber?.selectedOptionKeys]);
-                  onRuleNumberChange(
-                    ruleNumber?.selectedOptionKeys ? ruleNumber?.selectedOptionKeys : undefined
+                onChange={(changedRuleNumbers) => {
+                  setRuleNumber(
+                    changedRuleNumbers?.selectedOptionKeys
+                      ? changedRuleNumbers?.selectedOptionKeys
+                      : undefined
```

Comment:
Have you checked how frequently these components re-render? Something I'm a little concerned with is the fact that we're replacing local state that only affected a single component, like this one with the rule number, and now we're putting it into context. Wouldn't the context hook, and the rest of siblings re-render as soon as this value is mutated?

Reply:
Glad you mentioned this, the other components do in fact need to re-render.  Changing the rule filter needs to update the rules data fetched for the rules_table and rules_counters components.

If you pull the lastest `main` branch and turn on highlight update in the DevTools Profiler you will see many updates even outside of the Rules page.

A change like this is not ONLY about performance and re-renders,  this is making the logic easier to understand and work with.

![Screenshot 2024-11-08 at 12 57 58 PM](https://github.com/user-attachments/assets/2151c744-8bf9-4cb2-9a65-77a43f5f921a)